### ISSUE TITLE
Is there Kryo can support variable fields in object?

### ISSUE THREAD

**Issue by elkan1788** [EsotericSoftware/kryo#413]
Hi man. There are found a problem about variable fields, how can I fix it. The case like blow:

I had a Object need serialize, just name Food.

```
public class Food {
   private  int id;
   private  String name;
   ..........
}
```

But one day I find it not enough, so I extend some new fields. Now it look like this:

```
public class Food {
   private  int id;
   private  String name;
   private  double price;
   private  int counts;
   ..........
}
```

Now the problem is appear, when I deserialize the Food object, there had exist `Old Food` and `New Food`, so Kryo throw an exception **`kryo serialization failed buffer overflow`** .

So is there support this scene, the variable fields in object.

**Comment by elkan1788** [comment]
It was just found remove fields in [CompatibleFieldSerializerTest](https://github.com/EsotericSoftware/kryo/blob/master/test/com/esotericsoftware/kryo/CompatibleFieldSerializerTest.java) and [FieldSerializerTest](https://github.com/EsotericSoftware/kryo/blob/master/test/com/esotericsoftware/kryo/FieldSerializerTest.java), so how can I do!

**Comment by magro** [comment]
Please read the documentation: https://github.com/EsotericSoftware/kryo#compatibility
What's missing there? 

Btw, to ask questions please use the mailing list.

**Comment by elkan1788** [comment]
Thanks for answer. There can not visit the mailing list URL in **China**. I'm sorry. There are my first time use Kryo. Also it use to serialize and deserialize in Kafka message. I had read your document and test source but still can not understand how to use it. There are my code example:

```
public  class Food {
        private int id;
        private String name;

        // extend fields
        private double price;
        private int counts;
        ........
}
``` 

The test method like this:

```
@Test
    public void testCompatibleFieldSerializer2() throws Exception {

        Food food = new Food();
        food.setId(007);
        food.setName("rice");

        food.setPrice(2.00);
        food.setCounts(1);

        Kryo kryo = new Kryo();
        kryo.setDefaultSerializer(CompatibleFieldSerializer.class);
        kryo.register(Food.class);

        Output output = new Output(new ByteArrayOutputStream());
        kryo.writeObject(output, food);
        output.close();

        // store the serialize bytes on disk
        db.push(output.toBytes());

        System.out.println("=========================");
        // fetch the serialize bytes on disk, there will include old and new message.
        Map<Long, byte[]> data = db.pullAll();
        for (byte[] d : data.values()) {
            Input input = new Input(new ByteArrayInputStream(d));
            Food object = kryo.readObject(input, Food.class);
            input.close();
            System.out.println(object.toString());
        }
    }
```

So that is ` Food object = kryo.readObject(input, Food.class);` this line will throws exception like:

```
com.esotericsoftware.kryo.KryoException: Buffer underflow.
```

Is there any wrong code in my test???

**Comment by magro** [comment]
Ok, I didn't know that you cannot access google groups, sorry to hear that!

I cannot reproduce your issue. Instead of kafka I used `Output output = new Output(new java.io.FileOutputStream(file));` and serialized the first version of `Food` (with only `id` and `name` fields) to a file. Then I added two fields (`price` and `counts`) to `Food` and deserialized the previously writte file using `Input input = new Input(new FileInputStream(file));`. This was successful, the deserialized food instance was deserialized and printed as expected.
